tensorflow_graphics/projects/mesh_rcnn/ops/cubify.py

```diff
@@ -0,0 +1,303 @@
+  linear_index = tf.cast(tf.where(tf.not_equal(faces_idx, 0)), tf.int32)
+  nyxz = tf.transpose(tf.unravel_index(linear_index[:, 0],
+                                       (batch_size, height, width, depth)))
+
+  if len(nyxz) == 0:
+    return Meshes([], [])
+
+  unit_cube_faces_per_occupied_voxel = tf.gather(unit_cube_faces,
+                                                 linear_index[:, 1],
+                                                 axis=None)
+
+  def _compute_grid_coordinates(face_index):
+    xyz = tf.gather(unit_cube_verts,
+                    face_index,
+                    axis=None)
+    permute_idx = tf.constant([1, 0, 2])
+    yxz = tf.gather(xyz, permute_idx, axis=1)
+    yxz += nyxz[:, 1:]
+    return _ravel_index(yxz, (height + 1, width + 1, depth + 1))
+
+  grid_faces = tf.vectorized_map(_compute_grid_coordinates,
+                                 tf.transpose(
+                                     unit_cube_faces_per_occupied_voxel))
+
+  grid_faces = tf.transpose(grid_faces)
+
+  x, y, z = tf.meshgrid(tf.range(width + 1),
+                        tf.range(height + 1),
+                        tf.range(depth + 1))
+
+  # align the top left corner of each cuboid to the pixel coordinate of the
+  # input grid.
+  x = x * 2 / (width - 1) - 1.0
+  y = y * 2 / (height - 1) - 1.0
+  z = z * 2 / (depth - 1) - 1.0
+
+  grid_vertices = tf.reshape(tf.stack((x, y, z), axis=3), shape=(-1, 3))
+
+  # prepare filtering / merge of vertices.
+  n_vertices = grid_vertices.shape[0]
+  grid_faces += tf.reshape(nyxz[:, 0], shape=(-1, 1)) * n_vertices
+  idle_vertices = tf.ones(n_vertices * batch_size, dtype=tf.int32)
+  mask_indices, _ = tf.unique(tf.reshape(grid_faces, -1))
+  mask_indices = tf.expand_dims(mask_indices, -1)
+  idle_vertices = tf.tensor_scatter_nd_update(
+      idle_vertices,
+      mask_indices,
+      tf.zeros((mask_indices.shape[0],), dtype=tf.int32)
+  )
+  grid_faces -= tf.reshape(nyxz[:, 0], shape=(-1, 1)) * n_vertices
+  split_size = tf.math.bincount(nyxz[:, 0], minlength=batch_size)
+  faces_list = list(tf.split(grid_faces, split_size.numpy().tolist(), 0))
+  idle_vertices = tf.reshape(idle_vertices, (batch_size, n_vertices))
+  idle_n = tf.math.cumsum(idle_vertices, axis=1)
+
+  verts_list = [
+      tf.gather(grid_vertices,
+                tf.where((idle_vertices[n] == 0))[:, 0], axis=0) for n in
+      range(batch_size)
+  ]
+
+  faces_list = [face_batch - tf.gather(idle_n[n], face_batch, axis=0) for
+                n, face_batch in enumerate(faces_list)]
+
+  return Meshes(verts_list, faces_list, batch_sizes)
+
+
+def _create_face_mask(voxel_occupancy_grid, kernel, axis):
+  """
+  Creates face masks along one axis of a voxel occupancy grid. The
+  boundaries of the represented 3D shape are computed using 3D convolutions.
+
+  Example:
+    Consider a 1x2x2 fully occupied voxel grid. After applying this function for
+    dimension -1 (x coordinates), one receives two float32 tensors indicating
```

Comment:
voxel grid in ZYX coordinate ordering. 

This will help clarify x dim = -1, y-dim=-2, z = -3.

Why not use 0, 1, 2 instead of -1, -2, -3 as those are easier to understand.